File: main/src/main/scala/org/clulab/dynet/RowReaders.scala
/**
 * Task-specific readers for the Row class produced by ColumnReader
 * @author Mihai
 */

package org.clulab.dynet

import org.clulab.scala.WrappedArray._
import org.clulab.scala.WrappedArrayBuffer._
import org.clulab.sequences.Row
import org.clulab.utils.{Buffer, MathUtils}

import scala.collection.mutable
import scala.util.Random
import MetalRowReader._

case class AnnotatedSentence(words: IndexedSeq[String],
                             posTags: Option[IndexedSeq[String]] = None,
                             neTags: Option[IndexedSeq[String]] = None) {
  def indices: Range = words.indices
  def size: Int = words.size
}

trait RowReader {
  /** Converts the tabular format into one or more (AnnotatedSentence, sequence of gold heads (optional), sequence of gold labels) pairs */
  def toAnnotatedSentences(rows: IndexedSeq[Row], insertNegatives: Int = 0): IndexedSeq[(AnnotatedSentence, IndexedSeq[Label])]
}

class MetalRowReader extends RowReader {
  override def toAnnotatedSentences(rows: IndexedSeq[Row], insertNegatives: Int = 0): IndexedSeq[(AnnotatedSentence, IndexedSeq[Label])] = {

    if (rows.head.length == 2) {
      parseSimple(rows)
    } else if (rows.head.length == 4) {
      parseSimpleExtended(rows)
    } else if (rows.head.length >= 5) {
      parseFull(rows, insertNegatives)
    } else {
      throw new RuntimeException("ERROR: the Metal format expects 2, 4, or 5+ columns!")
    }
  }

  /** Parser for the simple format: word, label */
  def parseSimple(rows: IndexedSeq[Row]): IndexedSeq[(AnnotatedSentence, IndexedSeq[Label])] = {
    assert(rows.head.length == 2)
    val words = rows.map(_.get(WORD_POSITION))
    val labels = rows.map(row => PrimalLabel(row.get(WORD_POSITION + 1)))

    IndexedSeq((AnnotatedSentence(words), labels))
  }

  /** Parser for the simple extended format: word, POS tag, NE label, label */
  def parseSimpleExtended(rows: IndexedSeq[Row]): IndexedSeq[(AnnotatedSentence, IndexedSeq[Label])] = {
    assert(rows.head.length == 4)
    val words = rows.map(_.get(WORD_POSITION))
    val posTags = rows.map(_.get(POS_TAG_POSITION))
    val neLabels = rows.map(_.get(NE_LABEL_POSITION))
    val labels = rows.map(row => PrimalLabel(row.get(LABEL_START_OFFSET)))

    IndexedSeq(Tuple2(AnnotatedSentence(words, Some(posTags), Some(neLabels)), labels))
  }

  /** Parser for the full format: word, POS tag, NE label, (label head)+ */
  def parseFull(rows: IndexedSeq[Row], insertNegatives: Int): IndexedSeq[(AnnotatedSentence, IndexedSeq[Label])] = {
    assert(rows.head.length >= 5)
    val numSent = (rows.head.length - 3) / 2
    assert(numSent >= 1)

    val words = rows.map(_.get(WORD_POSITION))
    val posTags = rows.map(_.get(POS_TAG_POSITION))
    val neLabels = rows.map(_.get(NE_LABEL_POSITION))
    val labels = IndexedSeq.tabulate(numSent) { index =>
      rows.map(_.get(LABEL_START_OFFSET + (index * 2)))
    }
    val headPositions = IndexedSeq.tabulate(numSent) { index =>
      rows.map { row =>
        try {
          row.get(LABEL_START_OFFSET + (index * 2) + 1).toInt
        }
        catch {
          case _: NumberFormatException => -1
        }
      }
    }

    val annotatedSent = AnnotatedSentence(
      words,
      Some(posTags),
      Some(neLabels)
    )
    val sentences = labels.zip(headPositions).map { case (labelsForThisSentence, headsForThisSentence) =>
      val sentLabels = Buffer.makeArray[Label] { sentLabelsBuffer =>
        for (j <- labelsForThisSentence.indices) {
          sentLabelsBuffer += DualLabel(j, headsForThisSentence(j), labelsForThisSentence(j))

          if (insertNegatives > 0) {
            val negHeads = mkRandoms(-1 until annotatedSent.size, Set(headsForThisSentence(j)), insertNegatives)
            for (negHead <- negHeads) {
              sentLabelsBuffer += DualLabel(j, negHead, Utils.STOP_TAG)
            }
          }
        }
      }

      (annotatedSent, sentLabels: IndexedSeq[Label])
    }

    sentences
  }
}

object MetalRowReader {
  val WORD_POSITION = 0
  val POS_TAG_POSITION = 1
  val NE_LABEL_POSITION = 2
  val LABEL_START_OFFSET = 3

  val rand = new Random(1)

  private def mkRandoms(range: Range, exclude: Set[Int], howMany:Int): Set[Int] = {
    val numbers = MathUtils.randomize(range.toArray, rand)
    val randoms = new mutable.HashSet[Int]()
    for(n <- numbers if randoms.size < howMany) {
      if(! exclude.contains(n)) {
        randoms += n
      }
    }
    randoms.toSet
  }
}

File: main/src/main/python/sequences/rowReaders.py

class AnnotatedSentence:

    def __init__(self, words, posTags = None, neTags = None, headPositions = None):
        self.words = words
        self.posTags = posTags
        self.neTags = neTags
        self.headPositions = headPositions
        self.size = len(words)
        self.indicies = range(self.size)

class RowReader(object):

    def __init__(self):
        raise NotImplementedError

    def toAnnotatedSentences(self, rows):
        raise NotImplementedError

class MetalRowReader(RowReader):

    def __init__(self):
        self.WORD_POSITION = 0
        self.POS_TAG_POSITION = 1
        self.NE_LABEL_POSITION = 2
        self.LABEL_START_OFFSET = 3

    def toAnnotatedSentences(self, rows):
        if (rows[0].length == 2):
            return self.parseSimple(rows)
        elif (rows[0].length == 4):
            return self.parseSimpleExtended(rows)
        elif (rows[0].length >= 5):
            return self.parseFull(rows)
        else:
            raise RuntimeError("ERROR: the Metal format expects 2, 4, or 5+ columns!")

    # Parser for the simple format: word, label 
    def parseSimple(self, rows):
        assert(rows[0].length == 2)
        words = list()
        labels = list()

        for row in rows:
            words += [row.get(self.WORD_POSITION)]
            labels += [row.get(self.WORD_POSITION + 1)]

        return [(AnnotatedSentence(words), labels)]

    # Parser for the simple extended format: word, POS tag, NE label, label
    def parseSimpleExtended(self, rows):
        assert(rows[0].length == 4)
        words = list()
        posTags = list()
        neLabels = list()
        labels = list()

        for row in rows:
            words += [row.get(self.WORD_POSITION)]
            posTags += [row.get(self.POS_TAG_POSITION)]
            neLabels += [row.get(self.NE_LABEL_POSITION)]
            labels += [row.get(self.LABEL_START_OFFSET)]

        return [(AnnotatedSentence(words, posTags, neLabels), labels)]

    # Parser for the full format: word, POS tag, NE label, (label head)+ 
    def parseFull(self, rows):
        assert(rows[0].length >= 5)
        numSent = (rows[0].length - 3) / 2
        assert(numSent >= 1)
        assert(numSent==int(numSent))
        numSent = int(numSent)



        words = list()
        posTags = list()
        neLabels = list()
        headPositions = [list() for i in range(numSent)]
        labels = [list() for i in range(numSent)]

        for row in rows:
            words += [row.get(self.WORD_POSITION)]
            posTags += [row.get(self.POS_TAG_POSITION)]
            neLabels += [row.get(self.NE_LABEL_POSITION)]

            for j in range(numSent):
                labels[j]+= [row.get(self.LABEL_START_OFFSET + (j * 2))]
                try:
                    headPositions[j] += [int(row.get(self.LABEL_START_OFFSET + (j * 2) + 1))]
                except:
                    raise RuntimeError 

        sentences = list()
        for i in range(numSent):
            annotatedSent = AnnotatedSentence(words, posTags, neLabels, headPositions[i])
            sentLabels = labels[i]
            sentences += [(annotatedSent, sentLabels)]

        return sentences
